lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+
+  auto args = type->getGenericArgs();
+  if (args.size() != 1)
+    // Must've got something other than *Pointer or SIMD*
+    return clang::QualType();
+  auto argCanonicalTy = args[0]->getCanonicalType();
+
+  switch (kind) {
+  case StructKind::Invalid:
+    return clang::QualType();
+
+  case StructKind::UnsafeMutablePointer:
+  case StructKind::Unmanaged:
+  case StructKind::AutoreleasingUnsafeMutablePointer: {
+    auto clangTy = convert(argCanonicalTy);
+    if (clangTy.isNull())
+      return clang::QualType();
+    return ClangASTContext.getPointerType(clangTy);
+  }
+  case StructKind::UnsafePointer: {
+    return ClangASTContext.getPointerType(convert(argCanonicalTy).withConst());
+  }
+
+  case StructKind::CFunctionPointer: {
+    auto &clangCtx = ClangASTContext;
+
+    clang::QualType functionTy;
+    if (isa<SILFunctionType>(argCanonicalTy->getCanonicalType())) {
+      functionTy = convert(argCanonicalTy);
+    } else {
+      // Fall back to void().
+      functionTy = clangCtx.getFunctionNoProtoType(clangCtx.VoidTy);
+    }
+    return clangCtx.getPointerType(functionTy);
+  }
+
+  case StructKind::SIMD: {
+    clang::QualType scalarTy = convert(argCanonicalTy);
+    auto numEltsString = swiftStructDecl->getName().str();
+    numEltsString.consume_front("SIMD");
+    unsigned numElts;
+    bool failedParse = numEltsString.getAsInteger<unsigned>(10, numElts);
+    if (failedParse)
+      return clang::QualType();
+    (void) failedParse;
+    auto vectorTy = ClangASTContext.getVectorType(scalarTy, numElts,
+      clang::VectorType::VectorKind::GenericVector);
+    return vectorTy;
+  }
+  }
+
+  llvm_unreachable("Not a valid StructKind.");
+}
+
+clang::QualType ClangTypeConverter::visitEnumType(EnumType *type) {
+  // Special case: Uninhabited enums are not @objc, so we don't
+  // know what to do below, but we can just convert to 'void'.
+  if (type->isUninhabited())
+    return convert(Context.TheEmptyTupleType);
+
+  if (!type->getDecl()->isObjC())
+    // Can't translate something not marked with @objc
+    return clang::QualType();
+
+  // @objc enums lower to their raw types.
+  return convert(type->getDecl()->getRawType());
+}
+
+clang::QualType ClangTypeConverter::visitFunctionType(FunctionType *type) {
+  // We must've already computed it before if applicable.
+  return clang::QualType(type->getClangFunctionType(), 0);
+}
+
+clang::QualType ClangTypeConverter::visitSILFunctionType(SILFunctionType *type) {
+  llvm::report_fatal_error("Expected only AST types but found a SIL function.");
+}
+
+clang::QualType
+ClangTypeConverter::visitSILBlockStorageType(SILBlockStorageType *type) {
+  llvm::report_fatal_error("Expected only AST types but found a SIL block.");
+}
+
+clang::QualType
+ClangTypeConverter::visitProtocolCompositionType(ProtocolCompositionType *type) {
+  // Any will be lowered to AnyObject, so we return the same result.
+  if (type->isAny())
+    return getClangIdType(ClangASTContext);
```

Comment:
I'd forgotten that we're doing both bridging and translation in one step here.  Okay.

Reply:
I don't understand... aren't bridging and translation synonymous? What's the difference?

Reply:
The Swift type `Swift.Any` is bridged to the Swift type `Swift.AnyObject`, which can be translated to the Clang type `id`, which it is assumed to have identical ABI treatment to.